File: 140_word_break_2.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> List[str]:
        N = len(s)
        # ending[i] contains the ending indices in `s` of
        # words in `wordDict` that are prefixes of s[i:]
        ending = [[] for i in range(N)]
        reachable = [False] * (N + 1)
        reachable[0] = True
        for i in range(N):
            if reachable[i]:
                substr = s[i:]
                for w in wordDict:
                    if substr.startswith(w):
                        ending[i].append(i + len(w))
                        reachable[i + len(w)] = True
        if not reachable[-1]:
            return []
        sentence = {}
        for i, E in enumerate(ending):
            sen = sentence.get(i, [])
            if sen:
                for j in E:
                    sentence.setdefault(j, []).extend(x + ' ' + s[i:j] for x in sen)
                del sentence[i]
            else:
                for j in E:
                    sentence.setdefault(j, []).append(s[i:j])
        return sentence.get(len(s), [])




from collections import OrderedDict

class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> List[str]:
        N = len(s)
        # ending[i] contains the ending indices in `s` of
        # words in `wordDict` that are prefixes of s[i:]
        ending = OrderedDict()
        reachable = [True] + [False] * N
        for i in range(N):
            if reachable[i]:
                substr = s[i:]
                for w in wordDict:
                    if substr.startswith(w):
                        ending.setdefault(i, []).append(i + len(w))
                        reachable[i + len(w)] = True
        if not reachable[-1]:
            return []
        # sentences[i] holds all possible sentences of
        # words in s[:i]
        sentences = {}
        for i, E in ending.items():
            sen = sentences.get(i, [])
            if sen:
                for j in E:
                    sentences.setdefault(j, []).extend(x + ' ' + s[i:j] for x in sen)
                sentences.pop(i)
            else:
                for j in E:
                    sentences.setdefault(j, []).append(s[i:j])
        return sentences.get(len(s), [])

File: 140_word_break_2.cpp
#include <algorithm>
#include <map>


class Solution {
public:
    vector<string> wordBreak(string s, vector<string>& wordDict) {
        if (wordDict.empty())
            return {};
        vector<bool> reachable(s.size() + 1, false);
        reachable[0] = true;
        vector<vector<int>> ending(s.size() + 1);
        size_t maxlen = maxWordLength(wordDict);
        
        for (int i = 0; i < s.size(); i++){
            if (not reachable[i])
                continue;
            for (int j = 1; j <= min(maxlen, s.size() - i); j++){
                auto it = find(wordDict.begin(), wordDict.end(), s.substr(i, j));
                if (it != wordDict.end()){
                    reachable[i + j] = true;
                    ending[i].push_back(i + j);
                }
            }
        }
        if (!reachable.back())
            return {};
        map<int, vector<string>> sentences;
        for (int i = 0; i < ending.size(); i++){
            if (ending[i].empty())
                continue;
            
            auto it = sentences.find(i);
            if (it == sentences.end()){
                for (int j : ending[i])
                    sentences[i + j].push_back(s.substr(i, j - i));
            } else {
                for (const string& sen : it->second){
                    for (int j : ending[i])
                        sentences[j].push_back(sen + " " + s.substr(i, j - i));
                }
            }
        }
        return sentences[s.size()];
    }
    
    size_t maxWordLength(vector<string>& words){
        auto it = max_element(
            words.begin(), words.end(),
            [](const auto& x, const auto& y){ return x.size() < y.size();}
        );
        return it == words.end() ? 0 : it->size();
    }
};

